### PR TITLE
Update `azurerm_policy_set_definition` - Fix #8663

### LINKED ISSUE #8663: New issues introduced by #8270
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```shell
Terraform v0.13.3
+ provider registry.terraform.io/hashicorp/azurerm v2.29.0
```

### Affected Resource(s)

* `azurerm_policy_set_definition`

### Terraform Configuration Files

The code below is used to create multiple Policy Set Definitions from a `list` object  `local.es_policy_set_definitions_by_management_group`. This list is generated dynamically 'per Management Group' based on a map of Policy Set Definitions to assign per Management Group. The Policy Set Definitions are written in native ARM and stored as `.json` files within our module, and loaded into Terraform using the `file()` and `jsondecode()` functions.

A typical `local.es_policy_set_definitions_by_management_group` object looks like this:

```hcl
[
  {
    resource_id = <calculated_resource_id_of_resource_to_create>
    scope_id    = <resource_id_of_the_target_management_group>
    template    = <hcl_object_representation_of_policy_set_definition_from_template_file>
  }
]
```

The code used to deploy the Policy Set Definitions is as follows:

```hcl
resource "azurerm_policy_set_definition" "enterprise_scale" {
  for_each = {
    for policy_set in local.es_policy_set_definitions_by_management_group :
    policy_set.resource_id => policy_set
  }

  # Mandatory resource attributes
  name         = each.value.template.displayName
  policy_type  = "Custom"
  display_name = each.value.template.displayName

  # Dynamic configuration blocks
  dynamic "policy_definition_reference" {
    for_each = [
      for item in each.value.template.policyDefinitions :
      {
        policyDefinitionId          = item.policyDefinitionId
        parameters                  = item.parameters
        policyDefinitionReferenceId = item.policyDefinitionReferenceId
      }
      if try(length(each.value.template.policyDefinitions) > 0, false)
    ]
    content {
      policy_definition_id = policy_definition_reference.value["policyDefinitionId"]
      parameter_values     = try(length(policy_definition_reference.value["parameters"]) > 0, false) ? jsonencode(policy_definition_reference.value["parameters"]) : jsonencode(local.empty_map)
      reference_id         = try(length(policy_definition_reference.value["policyDefinitionReferenceId"]) > 0, false) ? policy_definition_reference.value["policyDefinitionReferenceId"] : policy_definition_reference.value["policyDefinitionId"]
    }
  }

  # Optional resource attributes
  description           = try(length(each.value.template.description) > 0, false) ? each.value.template.description : "${each.value.template.displayName} Policy Set Definition at scope ${each.value.scope_id}"
  management_group_name = try(length(each.value.scope_id) > 0, false) ? basename(each.value.scope_id) : null
  metadata              = try(length(each.value.template.metadata) > 0, false) ? jsonencode(each.value.template.metadata) : local.empty_string
  parameters            = try(length(each.value.template.parameters) > 0, false) ? jsonencode(each.value.template.parameters) : local.empty_string

  # Set explicit dependency on Policy Definition deployments
  depends_on = [
    azurerm_policy_definition.enterprise_scale,
  ]

}

```

Much of the logic is to handle situations where a source Policy Set Definition template may not contain certain optional values, and we use a `dynamic block` to generate the `policy_definition_reference` block based on the template content.

Following the updates released in v2.29.0 we have discovered a couple of issues which appear to be related as described below.

### Issue Description

Below is an extract of our code before #8270:

```hcl
  # Dynamic configuration blocks
  dynamic "policy_definition_reference" {
    for_each = [
      for item in each.value.template.policyDefinitions :
      {
        policyDefinitionId          = item.policyDefinitionId
        parameters                  = item.parameters
        policyDefinitionReferenceId = item.policyDefinitionReferenceId
      }
      if try(length(each.value.template.policyDefinitions) > 0, false)
    ]
    content {
      policy_definition_id = policy_definition_reference.value["policyDefinitionId"]
      parameters = {
        for parameter_id, parameter_value in policy_definition_reference.value["parameters"] :
        parameter_id => parameter_value.value
      }
      reference_id = try(length(policy_definition_reference.value["policyDefinitionReferenceId"]) > 0, false) ? policy_definition_reference.value["policyDefinitionReferenceId"] : policy_definition_reference.value["policyDefinitionId"]
    }
  }
```

Upon release of v2.29.0 we started to receive the following error:

```shell
Error: expanding `policy_definition_reference`: unmarshalling `parameter_values`: unexpected end of JSON input

  on modules/enterprise_scale/resources.policy_set_definitions.tf line 1, in resource "azurerm_policy_set_definition" "enterprise_scale":
   1: resource "azurerm_policy_set_definition" "enterprise_scale" {


Error: expanding `policy_definition_reference`: unmarshalling `parameter_values`: unexpected end of JSON input

  on modules/enterprise_scale/resources.policy_set_definitions.tf line 1, in resource "azurerm_policy_set_definition" "enterprise_scale":
   1: resource "azurerm_policy_set_definition" "enterprise_scale" {


Error: expanding `policy_definition_reference`: unmarshalling `parameter_values`: unexpected end of JSON input

  on modules/enterprise_scale/resources.policy_set_definitions.tf line 1, in resource "azurerm_policy_set_definition" "enterprise_scale":
   1: resource "azurerm_policy_set_definition" "enterprise_scale" {
```

We were able to narrow this down to the fact that since the changes introduced in #8270, the provider appears to no longer accept `null` for optional `parameter` values. This usually isn't an issue and is a pattern we use in multiple places within our code.

To fix this, we moved to the newly recommended parameter_values format which is actually beneficial to us as it simplifies the code as follows:

```hcl
  # Dynamic configuration blocks
  dynamic "policy_definition_reference" {
    for_each = [
      for item in each.value.template.policyDefinitions :
      {
        policyDefinitionId          = item.policyDefinitionId
        parameters                  = item.parameters
        policyDefinitionReferenceId = item.policyDefinitionReferenceId
      }
      if try(length(each.value.template.policyDefinitions) > 0, false)
    ]
    content {
      policy_definition_id = policy_definition_reference.value["policyDefinitionId"]
      parameter_values     = try(length(policy_definition_reference.value["parameters"]) > 0, false) ? jsonencode(policy_definition_reference.value["parameters"]) : jsonencode(local.empty_map)
      reference_id         = try(length(policy_definition_reference.value["policyDefinitionReferenceId"]) > 0, false) ? policy_definition_reference.value["policyDefinitionReferenceId"] : policy_definition_reference.value["policyDefinitionId"]
    }
  }

```

When no parameter value is provided in the original ARM template, we simply replace is with an empty JSON string `"{}"` which is generated using the syntax `jsonencode(local.empty_map)`, where `locals { empty_map = {} }`.

This initially appears to resolve the issue we were facing, however we've observed the following issues.

#### Issue 1
When trying to run `terraform apply` for the first time after the code update, Terraform wanted to update our Policy Set Definitions, but upon doing so we were faced with the following error:

```shell
Error: expanding `policy_definition_reference`: cannot set both `parameters` and `parameter_values`

  on .terraform/modules/enterprise_scale/resources.policy_set_definitions.tf line 1, in resource "azurerm_policy_set_definition" "enterprise_scale":
   1: resource "azurerm_policy_set_definition" "enterprise_scale" {

```

To work around this, we had to use `terraform taint` to force redeployment, at which point the deployment completed successfully.

#### Issue 2
Having got past the above issue, every time we try to run `terraform apply` we get prompted to update any Policy Set Definitions where no parameters are defined, or `"parameters": {}` and when apply is confirmed, we repeatedly get the same error:

```shell
Error: expanding `policy_definition_reference`: cannot set both `parameters` and `parameter_values`

  on .terraform/modules/enterprise_scale/resources.policy_set_definitions.tf line 1, in resource "azurerm_policy_set_definition" "enterprise_scale":
   1: resource "azurerm_policy_set_definition" "enterprise_scale" {

```

### Debug Output

Extract of debug at error point:

```shell
2020-09-28T19:24:24.804Z [DEBUG] plugin.terraform-provider-azurerm_v2.29.0_x5:
Id":"8251035221074005943","policyDefinitionId":"/providers/Microsoft.Management/managementGroups/tf/providers/Microsoft.Authorization/policyDefinitions/ES-Deny-PublicEndpoint-Aks","parameters":{}}]},"id":"/providers/Microsoft.Management/managementgroups/tf/providers/Microsoft.Authorization/policySetDefinitions/ES-Deny-Public-Endpoints-for-PaaS-Services","type":"Microsoft.Authorization/policySetDefinitions","name":"ES-Deny-Public-Endpoints-for-PaaS-Services"}
2020/09/28 19:24:24 [WARN] Provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new value for module.enterprise_scale.azurerm_policy_set_definition.enterprise_scale["/providers/Microsoft.Management/managementGroups/tf/providers/Microsoft.Authorization/policySetDefinitions/ES-Deny-Public-Endpoints-for-PaaS-Services"], but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .metadata: was cty.StringVal("{\"createdBy\":\"78665f88-b16d-473b-a073-9aaa5b6f6af3\",\"createdOn\":\"2020-09-28T12:06:47.1045376Z\",\"updatedBy\":\"78665f88-b16d-473b-a073-9aaa5b6f6af3\",\"updatedOn\":\"2020-09-28T19:17:57.8070067Z\"}"), but now cty.StringVal("{\"createdBy\":\"78665f88-b16d-473b-a073-9aaa5b6f6af3\",\"createdOn\":\"2020-09-28T12:06:47.1045376Z\",\"updatedBy\":\"78665f88-b16d-473b-a073-9aaa5b6f6af3\",\"updatedOn\":\"2020-09-28T19:22:53.6514379Z\"}")
      - .policy_definition_reference[0].parameter_values: was cty.StringVal("{}"), but now cty.StringVal("")
      - .policy_definition_reference[1].parameter_values: was cty.StringVal("{}"), but now cty.StringVal("")
      - .policy_definition_reference[2].parameter_values: was cty.StringVal("{}"), but now cty.StringVal("")
      - .policy_definition_reference[3].parameter_values: was cty.StringVal("{}"), but now cty.StringVal("")
      - .policy_definition_reference[4].parameter_values: was cty.StringVal("{}"), but now cty.StringVal("")
      - .policy_definition_reference[5].parameter_values: was cty.StringVal("{}"), but now cty.StringVal("")
      - .policy_definition_reference[6].parameter_values: was cty.StringVal("{}"), but now cty.StringVal("")
      - .policy_definition_reference[7].parameter_values: was cty.StringVal("{}"), but now cty.StringVal("")
```

If you can access our Terraform Cloud logs, the latest failed run with debug enabled is under ID `run-oLJ9fvae2xdzu4dt`.

I previously tested using just an empty string `""` as the default value, but this didn't work either.

### Panic Output

None.

### Expected Behavior

Should be able to run `terraform apply` multiple times and not have the apply update resources which haven't fundamentally changed, and complete without errors.

### Actual Behavior

As described above.

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

We are running this in standard public Azure, and are using Terraform Cloud for the remote backend (including deployment).

### References

* #0000

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes #8663

### CODE REVIEW

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @ArcturusZhang 
Thanks for this PR, just one comment below but otherwise LGTM 👍

File: azurerm/internal/services/policy/policy_set_definition_resource.go [hunk 1]
```diff
@@ -444,7 +444,7 @@ func expandAzureRMPolicySetDefinitionPolicyDefinitions(input []interface{}) (*[]
 		v := item.(map[string]interface{})
 
 		parameters := make(map[string]*policy.ParameterValuesValue)
-		if p, ok := v["parameter_values"]; ok {
+		if p, ok := v["parameter_values"]; ok && p.(string) != "" {
 			if err := json.Unmarshal([]byte(p.(string)), &parameters); err != nil {
 				return nil, fmt.Errorf("unmarshalling `parameter_values`: %+v", err)
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/policy/policy_set_definition_resource.go:449]: Done

**Comment by ArcturusZhang** [comment]
Hi @jackofallops I have made some changes, please have a look

File: azurerm/internal/services/policy/tests/policy_set_definition_resource_test.go [hunk 2]
```diff
@@ -619,6 +654,35 @@ PARAMETERS
...
+  policy_type  = "Custom"
+  mode         = "All"
+  display_name = "acctestpol-%d"
+
+  policy_rule = <<POLICY_RULE
+	{
+    "if": {
+      "not": {
+        "field": "location",
+        "equals": "eastus"
```
Comment (tombuildsstuff) [inline azurerm/internal/services/policy/tests/policy_set_definition_resource_test.go:674]: this won't work in regions other than Azure Public - so we need to pass this in here

**Review by tombuildsstuff — DISMISSED** [review]
hi @ArcturusZhang 

Thanks for pushing those changes - taking a look through there's another thing I've spotted here but if we can fix that up then this otherwise LGTM 👍 

Thanks!

**Review by jackofallops — APPROVED** [review]

**Comment by jackofallops** [comment]
Tests passing (2 failures are existing / unrelated)
image

**Pull request merged**